deep_qa/data/instance.py

```diff
@@ -0,0 +1,62 @@
+from typing import Dict, List
+
+import numpy
+
+from .fields import Field
+from . import Vocabulary
+
+
+class Instance:
+    """
+    An ``Instance`` is a collection of :class:`Field` objects, specifying the inputs and outputs to
+    some model.  We don't make a distinction between inputs and outputs here, though - all
+    operations are done on all fields, and when we return arrays, we return them as dictionaries
+    keyed by field name.  A model can then decide which fields it wants to use as inputs as which
+    as outputs.
+
+    The ``Fields`` in an ``Instance`` can start out either indexed or un-indexed.  During the data
+    processing pipeline, all fields will end up as ``IndexedFields``, and will then be converted
+    into padded arrays by a ``DataGenerator``.
+    """
+    def __init__(self, fields: Dict[str, Field]):
+        self._fields = fields
+
+    def count_vocab_items(self, counter: Dict[str, Dict[str, int]]):
+        """
+        Increments counts in the given ``counter`` for all of the vocabulary items in all of the
+        ``Fields`` in this ``Instance``.
+        """
+        for _, field in self._fields.items():
+            if field.needs_indexing():
+                field.count_vocab_items(counter)
+
+    def index_fields(self, vocab: Vocabulary):
+        """
+        Converts all ``UnindexedFields`` in this ``Instance`` to ``IndexedFields``, given the
+        ``Vocabulary``.  This `mutates` the current object, it does not return a new ``Instance``.
+        """
+        for _, field in self._fields.items():
+            if field.needs_indexing():
```

Comment:
An alternate is to just remove `needs_indexing` from the API and make `index` a no-op for those that don't need indexing.